### ISSUE TITLE
installDeps.sh: use "npm ci" instead of "npm install"?

### ISSUE THREAD

**Issue by muxator** [ether/etherpad-lite#3778]
In #3750 a user implicitly suggested we could use `npm ci` to install dependencies. At the moment, `bin/installDeps.sh` uses `npm install`:

https://github.com/ether/etherpad-lite/blob/1ad6e9c2880ed074ba7f6185bdb399d84b298310/bin/installDeps.sh#L103

Reading from the [`npm ci` docs](https://docs.npmjs.com/cli/ci.html):

> it’s meant to be used in [...] any situation where you want to make sure you’re doing a clean install of your dependencies.
> It can be **significantly faster** than a regular npm install by skipping certain user-oriented features.
> It is also more strict than a regular install, which **can help catch errors or inconsistencies caused by the incrementally-installed local environments** of most npm users.

## Possible advantages
1. **increased reproducibility** of the installations (both direct and via Docker);
2. once installed, **checking dependencies would be very fast**. This would solve the underlying issue below #3711.

   ~~Afterwards, we can still commit [`fastRun.sh`](https://github.com/ether/etherpad-lite/issues/3711#issuecomment-599135831) if we want.~~ **Edit**: `fastRun.sh` was commited in 8956efc4ae020bb75663be32b73de6fc8d36597b and will be part of **1.8.3**);
3. **often, user reported problems are due to poorly installed environments**. This could help reduce the confusion.

## Possible problems
1. **compatiblity with installing plugins** has to be checked.

   Plugins are internally installed via `npm install`, which would render inconsistent the contents of `node_modules` and `package-lock.json`.
   On the other side, plugins seem to end up in a different directory.

   This really has to be checked.

2. among other things, being more strict means that:

   > **If dependencies** in the package lock **do not match** those in package.json, **npm ci will exit with an error**, instead of updating the package lock.

   This **may confuse our users**, but it's better to have a clear error which a clear solution (== reinstall) than having half-backed installs laying around, or random bug report coming in;

3. something different I may be overlooking

**Comment by JohnMcLear** [comment]
+1, why not make the sauce labs runner use npm ci instead and test it that way?   That would be a good enough baptism of fire no?

**Comment by muxator** [comment]
> why not make the sauce labs runner use npm ci instead and test it that way?

Good idea!

Problem is `.travis.yml` calls `installDeps.sh`:
https://github.com/ether/etherpad-lite/blob/1ad6e9c2880ed074ba7f6185bdb399d84b298310/.travis.yml#L29

We should add an option or temporarily duplicate `installDeps.sh` as `installCI.sh`(that would kill me, because it then would live on for another 8 years).

**Comment by JohnMcLear** [comment]
No.  We could just create a PR on a branch w/ the change(to installDeps.sh) and travis will respect that....??







-----Original message-----

**Comment by muxator** [comment]
Yay.

**Comment by JohnMcLear** [comment]
See ci branch PR

**Comment by JohnMcLear** [comment]
``npm ERR! cipm can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.``

**Comment by JohnMcLear** [comment]
npm 6+ btw..  So I'm not sure this is a win for us..  You have to use npm install anyway first time around?

**Comment by muxator** [comment]
This is good.

I think this happened when you last updated the ueberdb dependency. After that change I started to find my dependencies out of sync as well. Now the CI catches it.

If I am right, it should only be a matter of upgrading the dependencies in the package-lock.json on your PR.

**Comment by muxator** [comment]
> You have to use npm install anyway first time around?

I guess no: it worked for me with an empty `node_modules`.

**Comment by JohnMcLear** [comment]
Travis shit itself too, installDeps failed...

**Comment by muxator** [comment]
Right away.

**Comment by muxator** [comment]
Did it work?

**Comment by JohnMcLear** [comment]
```
Error: Cannot find module '../../../src/node_modules/wd'

Require stack:

- /home/travis/build/ether/[secure]/tests/frontend/travis/remote_runner.js

    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)

    at Function.Module._load (internal/modules/cjs/loader.js:864:27)

    at Module.require (internal/modules/cjs/loader.js:1044:19)

    at require (internal/modules/cjs/helpers.js:77:18)

    at Object.<anonymous> (/home/travis/build/ether/[secure]/tests/frontend/travis/remote_runner.js:2:10)

    at Module._compile (internal/modules/cjs/loader.js:1158:30)

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)

    at Module.load (internal/modules/cjs/loader.js:1002:32)

    at Function.Module._load (internal/modules/cjs/loader.js:901:14)

    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12) {

  code: 'MODULE_NOT_FOUND',

  requireStack: [

    '/home/travis/build/ether/[secure]/tests/frontend/travis/remote_runner.js'

  ]

}

tests/frontend/travis/runner.sh: 14: kill: No such process

The command "tests/frontend/travis/runner.sh" exited with 1.
cache.2

store build cache

Done. Your build exited with 1.
```
Don't you get the Travis logs?  You need to see them :D

**Comment by muxator** [comment]
It did not install the dev dependencies.

> Don't you get the Travis logs? You need to see them :D

This one, right? https://travis-ci.org/github/ether/etherpad-lite/jobs/668225504

**Comment by JohnMcLear** [comment]
Ja perfect :D 

I guess maybe npm ci doesn't install dev deps?

**Comment by muxator** [comment]
It does on my machine:

```
.../etherpad-lite$ ls -1d src/node_modules/{mocha,nyc,supertest,wd}
src/node_modules/mocha
src/node_modules/nyc
src/node_modules/supertest
src/node_modules/wd
```

This is after deleting `node_modules` and running `bin/installDeps.sh` on the current `ci` branch.


Is it possible that there is a `NODE_ENV=production` somewhere in the CI environment? That would explain the missing dev dependencies. Just guessing.

**Comment by JohnMcLear** [comment]
99% no.  You could try and submit another commit to the branch and see if it was a one off bug, SL is inconsistent so you often have to run tests multiple times :)

**Comment by JohnMcLear** [comment]
``npm ci`` creates ``wd`` for me too ;\ let's try another commit?  See if it fixes things..

**Comment by muxator** [comment]
Committed 6d9d148651bb in the ci branch (I am going to delete that branch :) )

Let's see.

**Comment by JohnMcLear** [comment]
no dice ;(

**Comment by muxator** [comment]
Do you see https://travis-ci.org/github/ether/etherpad-lite/jobs/668231383#L203?
```
> wd@1.11.4 install /home/travis/build/ether/[secure]/src/node_modules/wd
```

It seems that, after all, `wd` is installed by Travis.
But then the subsequent errors are suspect.

I am trying to change `travis/runner.sh` (1924aab6b1c8) based on some feelings those errors gave me.

**Comment by muxator** [comment]
The error changed. Let's hope it's a good sign.

https://travis-ci.org/github/ether/etherpad-lite/jobs/668235282#L221

**Comment by muxator** [comment]
I think I nailed it.

https://travis-ci.org/github/ether/etherpad-lite/jobs/668241855

This version 25bd3016e439 (soon to be deleted) is the first one on which tests run.

I had to modify `tests/frontend/travis/runner.sh`. After looking at it & the Travis logs, I wonder how it worked before.

I'll leave everything as-is, and eventually port these modifications on the development branch. They are needed anyway.

**Comment by JohnMcLear** [comment]
Hah nicely done!  Though we should run a few tests because a lot of tests are failing due to sauce labs not executing them properly, I have an open support ticket.

**Comment by muxator** [comment]
I am merging #3789 in `develop`.

That PR introduces a new `travis/runner.sh` script which is compatible with both `npm install` & `npm ci`. There are other advantages (detailed in the PR).

**Comment by JohnMcLear** [comment]
@muxator this can now be closed?

**Comment by JohnMcLear** [comment]
Please re-open if I get this wrong, I'm 99% sure you merged ``ci`` logic.

**Comment by muxator** [comment]
This has to stay open.

The `npm ci` logic has not been integrated yet, because we still have to evaluate what I wrote in the proposal.

However, [we fixed `travis/runner.sh`](4a17443a2e6ad144faa6919f129359e8474b011d) that would  have broken as soon as we made any change.

**Comment by JohnMcLear** [comment]
Please link to the proposal.

**Comment by muxator** [comment]
> Please link to the proposal.

The proposal is at the start of this ticket. :smile:

This is definitely needed, but cannot be done in **1.8.3** because the plugin installation mechanism is complex to understand, and I do not want to risk to break it.

In the meantime, two more users got confused (#3911) by our startup procedure, or they could have benefitted from a clear error in case their dependencies are broken (#3916).

**Comment by JohnMcLear** [comment]
With ``ci`` set in installDeps.sh

## npm cli
```
rm -Rf node_modules/mammoth
rm -Rf src/node_modules
npm install ep_mammoth
bin/run.sh
```
WORKS

## git
```
rm -Rf node_modules/mammoth
rm -Rf node_modules/ep_mammoth
git clone git@github.com:JohnMcLear/ep_mammoth.git
rm -Rf node_modules/mammoth
rm -Rf src/node_modules
bin/run.sh
```
FAILS

```
[2020-04-23 18:37:07.961] [ERROR] console - Failed to load 'ep_mammoth/import.js' for 'ep_mammoth/ep_mammoth/hooks/import': Error: Cannot find module '../../node_modules/mammoth'
Require stack:
- /home/jose/develop/node_modules/ep_mammoth/import.js
- /home/jose/develop/src/static/js/pluginfw/shared.js
- /home/jose/develop/src/static/js/pluginfw/plugins.js
- /home/jose/develop/src/node/server.js
```

Changing from ``ci`` to ``install`` for line 103 in ``installDeps.sh``
```
rm -Rf node_modules/mammoth
rm -Rf node_modules/ep_mammoth
git clone git@github.com:JohnMcLear/ep_mammoth.git
rm -Rf node_modules/mammoth
rm -Rf src/node_modules
bin/run.sh
```

FAILS

```
[2020-04-23 18:37:07.961] [ERROR] console - Failed to load 'ep_mammoth/import.js' for 'ep_mammoth/ep_mammoth/hooks/import': Error: Cannot find module '../../node_modules/mammoth'
Require stack:
- /home/jose/develop/node_modules/ep_mammoth/import.js
- /home/jose/develop/src/static/js/pluginfw/shared.js
- /home/jose/develop/src/static/js/pluginfw/plugins.js
- /home/jose/develop/src/node/server.js

```

## Using WebUI
Installing plugin using the web ui works as expected.

# Conclusion

1. ``git clone`` without doing an ``npm install`` is never a reliable way to install a plugin that has any dependencies.

1. Using ``ci`` in installDeps.sh seems just as reliable as ``install`` as far as plugins go.

# Concerns

None.  IMHO ``ci`` is fine to be merged into 1.8.4

**Comment by muxator** [comment]
Fine, thanks for the assessment!

It will be included in **1.8.4**, then.
